### ISSUE TITLE
Some streams refuse to play on Vivaldi for Windows unless the user agent is spoofed

### ISSUE THREAD

**Issue by pieterdd** [eshaz/icecast-metadata-js#204]
I'm experiencing a fairly specific issue that manifests in the Windows version of Vivaldi but not on Linux. When I try to play some of the SomaFM streams from https://eshaz.github.io/icecast-metadata-js/, playback refuses to start:

![image](https://github.com/user-attachments/assets/42d67db8-e466-48b6-98e2-04e24f5a1a4b)

The peculiar thing is that when played separately, the stream runs fine:

![image](https://github.com/user-attachments/assets/034dece5-b896-4f97-b47d-97533d130195)

I've also tried embedding the stream into a webpage without relying on icecast-metadata-js, and it plays fine. Now here's where things get quite wild: when I fake my Vivaldi user agent, the stream plays without issue:

![image](https://github.com/user-attachments/assets/b5d068c0-ac0d-4ca4-9e02-6e1ba6e5a174)

Chrome on Windows doesn't have this issue, so it appears to be quite specific to Vivaldi. I searched the code of your repo before posting this, and it doesn't look like you have any conditional logic depending on the value of `navigator.userAgent`, so I'm kind of puzzled. Do you have any idea why this issue might manifest specifically in Vivaldi on Windows, and only if I don't spoof my user agent?

Thank you for making this library, by the way. You might be happy to know that the new SomaFM web player is using it :)

**Comment by eshaz** [comment]
Hello, thanks for reaching out. The user agent shouldn't matter at all for the client side library. It's possible that SomaFM's Icecast server is doing something odd with the user agent.

There is some logic in Icecast-kh that checks the user agent and does something (not sure exactly what). https://github.com/karlheyes/icecast-kh/blob/ea3554b5167f6780a8be77f1715f90ed02005665/src/format.c#L329 It looks like SomaFM is using Icecast 2.4.0-kh15 based on their `Server` response header

What does Vivaldi send as the user agent, and what did you change it to when it worked? I've never used Vivaldi, but I can install and try it out on Windows.

That's awesome that SomaFM is using my library, I had no idea! They are definitely one of the best internet radio stations out there.

**Comment by pieterdd** [comment]
Thank you for sharing your hypothesis. I've created https://jsfiddle.net/w8brjt92/ to test playing one of the affected streams through a vanilla <audio> tag:

```html
<audio src="https://ice1.somafm.com/beatblender-64-aac" controls autoplay />
```

This works for me in Vivaldi on Windows. The User-Agent value sent to SomaFM's server is `Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36`.

When I play the same stream through https://eshaz.github.io/icecast-metadata-js/, the User-Agent value is identical: `Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/130.0.0.0 Safari/537.36`. I'll check if there's any other difference in request headers.

**Comment by pieterdd** [comment]
I've created a three-way diff of the request headers.

- Left: https://eshaz.github.io/icecast-metadata-js/ without user agent spoofing (doesn't work)
- Middle: https://eshaz.github.io/icecast-metadata-js/ with user agent spoofed to `Mozilla/5.0 (Windows NT 6.1; WOW64; rv:33.0) Gecko/20120101 Firefox/33.0` (works here)
- Right: https://jsfiddle.net/w8brjt92/ without user agent spoofing (works here)

![image](https://github.com/user-attachments/assets/3023c027-cfbe-4234-89f1-2422b7444dd2)

Comparing left and middle, only the hostname and user agent appear to be different.

**Comment by rustyhodge** [comment]
Maybe this has something to do with the audio/aac vs audio/aacp headers?  Usually aac-he is audio/aacp and I bet that's what's breaking things.

**Comment by rustyhodge** [comment]
Nevermind, I tried changing the mimetype to audio/aac but that didn't fix it.

However in the constants.js you have:

`export const MP4A_40_2 = "mp4a.40.2";`

But not an entry for `mp4a.40.5` or `mp4a.40.29`

Could that be part of the problem?

**Comment by eshaz** [comment]
I installed Vivaldi and tested this. It looks like this is limited to AAC streams only.

Something interesting is that the Groove Salad AAC streams work, but other SomaFM AAC streams like Groove Salad Classic don't. This also isn't limited to just SomaFM streams, a few other AAC streams on the demo page have this issue. @rustyhodge, is there some difference in how these two servers / sources are configured? It appears Groove Salad AAC is being encoded at the source differently than the other streams.

I think I found the issue though. `IcecastMetadataPlayer` uses the MediaSource api as one of the methods of playing the audio after the ICY metadata has been parsed and removed from the response. Support for this api varies between browsers and host operating systems, and is sometimes implemented incorrectly. Vivaldi on Windows returns true here https://github.com/eshaz/icecast-metadata-js/blob/15cf791f30b4eb99f4beeebe2dec65c9154f2232/src/icecast-metadata-player/src/players/MediaSourcePlayer.js#L157-L163 which means the library will send the raw AAC data into the MediaSource api, but it's actually not able to decode some streams.

It is able to decode the audio when the AAC data is wrapped in an ISOBMFF container though (fMP4). I can probably switch around this logic to test compatibility with the containerized AAC data first, and then fallback to the raw AAC data if the container isn't supported. This should solve this problem without (hopefully) creating any new problems.

Also, @rustyhodge thanks for the donation, it's much appreciated!

**Comment by rustyhodge** [comment]
This is specifically limited to AAC-HE streams; AAC-LC work.

The following channels use an Orban AAC-LC encoder for 128k, and send a audio/aac mime-type header.

dronezone
groove salad
illstreet
indiepop
missioncontrol
secretagent
spacestation
suburbsofgoa
u80s

And these use the "sc_trans" encoder, which also use AAC-LC and sends a audio/aac mime-type header.

defcon
doomed
live
thetrip
scanner

All the rest use  Shoutcast AAC encoders, which encode 128k as AAC-HE and sends audio/aacp as the mime type.

examples are: xmastinfrisko, seventies, gsclassic, poptron, secretagent

Hope this helps.

**Comment by eshaz** [comment]
@pieterdd I've released icecast-metadata-player/1.17.4 that should solve this issue. Let me know if there are any further issues.

**Comment by eshaz** [comment]
@rustyhodge, would you consider adding Opus streams to SomaFM?

**Comment by pieterdd** [comment]
Thanks Ethan! I can no longer reproduce the issue with the new version, so everything's looking good here.

**Comment by rustyhodge** [comment]
I too confirm that the Vivaldi Windows version is fixed by this, and even renders correctly (vs Mac Vivaldi, which renders 32 AAC-HE as LC).

Thanks for your assistance with this!

**Comment by rustyhodge** [comment]
@eshaz could you summarize what had to be done to fix the issue? I think you're doing some further decoding of the stream to catch the AAC-HE frames and adjust the codec? (I suck at Javascript)

**Comment by rustyhodge** [comment]
@eshaz These changes unfortunately broke our player in Safari Mac. (iOS is unaffected) It now just hangs on AAC-LC and -HE content. You can test at https://somaplayer-preview.netlify.app/player/station/gsclassic

PS- https://eshaz.github.io/icecast-metadata-js/ won't display on a Mac either so I can't test on that.

**Comment by eshaz** [comment]
Ok, I'll take a look this afternoon. I don't have a Mac to test, I just use a VM, and it's a bit tedious. I'd recommend going back to the previous version while this issue gets sorted out.

**Comment by pieterdd** [comment]
https://somaplayer-preview.netlify.app/player/ has since rolled back, but https://deploy-preview-8--somaplayer-preview.netlify.app/player/ should still be using the updated version.

**Comment by eshaz** [comment]
Currently, all AAC gets sent to the MSE api as `mp4a.40.2`. Maybe this is causing an issue in Safari.

Looks like the AAC types mentioned above map to the below.
https://developer.mozilla.org/en-US/docs/Web/Media/Formats/codecs_parameter
| ISOBMFF codec | AAC type |
| --- | --- |
| `mp4a.40.2` | AAC LC (Low Complexity) |
| `mp4a.40.5` | HE-AAC, HE-AAC v2 |
| `mp4a.40.29` | HE-AAC v2 |

**Comment by eshaz** [comment]
@pieterdd, @rustyhodge, I've released a different approach that should work for both browsers in icecast-metadata-player/1.17.5

Both Safari (non-iOS) and Vivaldi support the [MediaSource](https://developer.mozilla.org/en-US/docs/Web/API/MediaSource) api for AAC playback, whcih is what IcecastMetadataPlayer will use by default. The issue is that Safari and Vivaldi are too optimistic when `MediaSource.isTypeSypported()` is called for AAC. They both return `true`, for `audio/aac` and `audio/mp3;codecs=mp4a.40.2`, but actually Safari only supports sending ADTS AAC data i.e. `audio/aac`, and Vivaldi only supports ISOBMFF wrapped AAC `audio/mp3;codecs=mp4a.40.2`. The MediaSource api on these browsers claim both codecs will work, but then throw an error when IceacstMetadataPlayer starts to send in the data. The previous solution just changed the order that the codec types are selected, which fixed Vivaldi but broke Safari.

The solution this time is to add failover logic so that when the decode error happens, the MediaSourcePlayer will try again using the next supported codec / wrapper. This also might help in other scenarios where fmp4 and webm are both "supported", but then one actually fails to decode.

I'm doing the muxing with another another library I wrote, [mse-audio-wrapper](https://github.com/eshaz/mse-audio-wrapper). In addition to AAC, it also allows MP3, Opus, Vorbis, and Flac playback through the MediaSource api.

I tested this change on Vivaldi (Windows), Firefox (Linux, Windows, Android), Chrome (Linux, Windows, Android), Safari (Mac Desktop, and iOS 13).

Let me know if something still isn't working.

> PS- https://eshaz.github.io/icecast-metadata-js/ won't display on a Mac either so I can't test on that.

@rustyhodge I was able to see and use the page on my MacOS VM on Safari. Did you see any errors when it wouldn't load for you?
